### PR TITLE
TestGetOptimalQuoteInGivenOut: Tests

### PR DESCRIPTION
**Pull request by deividaspetraitis**
This PR covers `TestGetOptimalQuoteInGivenOut` with tests.

As a side effect `GetOptimalQuote` method is broken is smaller parts to in order to be able to mock it. Such approach allows to tests smaller units indepently as well.

I have tested indroduced changes with Unit tests and integration tests:

```
============================================= short test summary info ==============================================
FAILED tests/test_candidate_routes.py::TestCandidateRoutes::test_all_valid_tokens[http://localhost:9092-ibc/0F9E9277
B61A78CB31014D541ACA5BF6AB06DFC4524C4C836490B131DAAECD78] - AssertionError: Error: no routes found for token in ibc/
0F9E9277B61A78CB31014D541ACA5BF6AB06DFC4524C4C836490B13...
FAILED tests/test_pools.py::TestPools::test_canonical_orderbook[http://localhost:9092] - AssertionError: Canonical o
rderbooks are None
FAILED tests/test_tokens_prices.py::TestTokensPrices::test_low_liq_token_prices[http://localhost:9092-ibc/56C276FC13
6E239449DCE664292DBEEF5795C4EF4B5B35DB98BD1C0948274867] - AssertionError: ibc/56C276FC136E239449DCE664292DBEEF5795C4
EF4B5B35DB98BD1C0948274867, SQS price is zero
==================================== 3 failed, 1490 passed, 8 skipped in 31.31s ====================================

```

This PR will be merged is `DATA-276` branch.

## Summary by CodeRabbit

- **New Features**
	- Introduced new parameters for token transactions, enhancing API flexibility.
	- Added methods for managing token denominations in various pools.
	- Implemented new quote retrieval methods for improved routing functionality.

- **Bug Fixes**
	- Refined error handling across the router, improving clarity and maintainability.

- **Documentation**
	- Updated API documentation to reflect changes in query parameters and error handling.

- **Tests**
	- Expanded test suite for the Router functionality, including new test cases for optimal quote retrieval.

- **Chores**
	- Enhanced code readability and organization across multiple files.

### CODE REVIEW

**Comment by deividaspetraitis** [comment]
Linter is complaining about following which stems from original `DATA-276` branch. I will resolve it there:

```
Error: `(*RouterTestHelper).newRoutablePool` - `cosmWasmConfig` always receives `domain.CosmWasmPoolRouterConfig{}` (`nil`) (unparam)
```

File: router/usecase/routertesting/quote.go [hunk 1]
```diff
@@ -52,10 +52,22 @@ func (s *RouterTestHelper) newRoutablePool(pool sqsdomain.PoolI, tokenOutDenom s
...
 
+// NewAmountIn creates a new amount in.
+func (s *RouterTestHelper) NewAmountIn() sdk.Coin {
+	return sdk.NewCoin(ETH, totalInAmount)
+}
+
+// NewAmountOut creates a new amount out.
+func (s *RouterTestHelper) NewAmountOut() osmomath.Int {
+	return totalOutAmount
+}
```
Comment (p0mvn) [inline router/usecase/routertesting/quote.go]: What is the purpose of these helpers? Consider exporting the variables and using them directly if needed

**Review by p0mvn — COMMENTED** [review]
Thanks for separating this into a separate PR for ease of review.

What do you think about setting up our "mainnet state mock" test suite that creates a mock test environment from mainnet state rather than wiring a mock for GetOptimalQuote?

I know some time has been spent on the GetOptimalQuote mock but I'm not sure how I'm feeling about it due to the added wiring complexity to the core codebase while covering a few branches

Similar to [this test](https://github.com/osmosis-labs/sqs/blob/5135a6f19360713ed7348c47f5afb9e50d74b8f2/router/usecase/optimized_routes_test.go#L555). Note [these lines](https://github.com/osmosis-labs/sqs/blob/5135a6f19360713ed7348c47f5afb9e50d74b8f2/router/usecase/optimized_routes_test.go#L650-L653) that set up a mainnet environment.

In fact, it seems that your test already leverages a [helper](https://github.com/osmosis-labs/sqs/blob/5135a6f19360713ed7348c47f5afb9e50d74b8f2/router/usecase/router_usecase_test.go#L698) that sets up a router use case with the mainnet state. With the approach I suggest, we will not be able to cover [this branch](https://github.com/osmosis-labs/sqs/blob/5135a6f19360713ed7348c47f5afb9e50d74b8f2/router/usecase/router_usecase.go#L289-L292). However, that is acceptable because we will benefit from deeper functional testing while reducing complexity from not having to monkey-patch GetOptimalQuote

Let me know what you think

**Comment by deividaspetraitis** [comment]
Thanks for taking your time to review and sharing thoughts!

While this PR touches core logic and changes wiring I believe it will support long term project testability strategy, goals and general code base maintainability:

- Other methods and funcs benefits from this change by allowing us to test and mock many other methods independently, both public for example `GetCachedRankedRoutes` and private `rankRoutesByDirectQuote`. 
- Testing specific part(s) while ignoring other parts would allow us to write more focused and indpedendent tests with no side effects
- Tests execution will be faster, since each test tests only specific code piece
- Decoupling allows wore isolated changes

For example, for `TestGetOptimalQuote_Mainnet` it makes sense as it is done currently, for `TestGetOptimalQuoteInGivenOut` method `GetOptimalQuote` is a depedency that is already cover with tests that's why I think points above makes more sense in this case.

Please let me know if you still think we should go for `TestGetOptimalQuote_Mainnet` approach and I will implement it! :pray:

File: router/usecase/routertesting/quote.go (hunk 1 above)
Comment (deividaspetraitis) [reply router/usecase/routertesting/quote.go]: Updated accordingly!

**Comment by p0mvn** [comment]
@deividaspetraitis let's brainstorm this change sync sometime next week and align on direction.

I think `TestGetOptimalQuote_Mainnet` would still be beneficial since it would be a functional test relying on mainnet state, asserting invariants more comprehensively.

I propose kicking off a separate PR with the mainnet state test.

Let me know what you think

**Comment by deividaspetraitis** [comment]
> @deividaspetraitis let's brainstorm this change sync sometime next week and align on direction.
> 
> I think `TestGetOptimalQuote_Mainnet` would still be beneficial since it would be a functional test relying on mainnet state, asserting invariants more comprehensively.
> 
> I propose kicking off a separate PR with the mainnet state test.
> 
> Let me know what you think

I have created separate PR with the mainnet state test https://github.com/osmosis-labs/sqs/pull/403 :rocket:

**Comment by deividaspetraitis** [comment]
While I still believe that merging suggestions introduced in this PR would be beneficial, however due working on more high priority tasks it would be hard to allocate time for further discussions, thus I am cllosing this for now. We can always get back to it in the future. cc @p0mvn

**Pull request closed without merging**